Add flags to dump IR to a file before and after LLVM passes

NuriAmari opened the pull request:
LLVM offers -print-after and -print-before flags that allow you to print IR to stderr before and after any pass you want. This can be useful for debugging LLVM optimization issue, but is far too noisy for large builds.

This patch adds analogous options -dump-after and -dump-before that dump the IR to appropriately named files. In addition, it also introduces flags -dump-after-all, -dump-before-all and -ir-dump-directory to control where the files are written to.

The implementation follows @chandlerc suggestions in https://reviews.llvm.org/D54794, but does not implement the suggested design completely. Notably, the IR extraction behavior differing from the print family of flags is not implemented. In addition to the suggestion of adding how many times a particular pass has run in at a given nesting level, I've also added how many passes have run at that nesting level. The tests provide a more clear explanation.

This is not complete, but I think it is an acceptable initial pass, and I'd like some feedback on the approach before I go too deep. Features that are missing / I would like:

- Less verbose pass names (particular pass managers and adapters have extremely verbose names)
- More descriptive names (ie. what function is this function pass operation on)
- Filter parity with the print flags (those flags respect flags for example allowing filtering of which functions to print)
- Legacy pass manager support (we need this to work for MIR, and at first glance it doesn't)

Commit:
Add flags to dump IR to a file before and after LLVM passes

Summary:

LLVM offers -print-after and -print-before flags that allow you
to print IR to stderr before and after any pass you want. This can
be useful for debugging LLVM optimization issue, but is far too
noisy for large builds.

This patch adds analogous options -dump-after and -dump-before that
dump the IR to appropriately named files. In addition, it also
introduces flags -dump-after-all, -dump-before-all and -ir-dump-directory
to control where the files are written to.

Test Plan:

Included LIT tests:
```
ninja check-llvm
```

NuriAmari commented on llvm/lib/IR/PrintPasses.cpp:
```diff
@@ -33,6 +33,31 @@ static cl::opt<bool> PrintAfterAll("print-after-all",
                                    llvm::cl::desc("Print IR after each pass"),
                                    cl::init(false), cl::Hidden);
 
+static cl::list<std::string>
```
There is a lot duplication between the print code and the dump code, though there are differences, it might be nice to try to combine them a little more.

NuriAmari commented on llvm/include/llvm/Passes/StandardInstrumentations.h:
```diff
@@ -69,6 +69,62 @@ class PrintIRInstrumentation {
+
+private:
+  void dumpBeforePass(StringRef PassID, Any IR);
+  void dumpAfterPass(StringRef PassID, Any IR);
```
I did not implement a `dumpAfterPassInvalidated` callback, mostly because I didn't understand when that occurs. If it makes sense to include one, I can.

NuriAmari commented on llvm/test/Other/dump-before-after-filenames:
```diff
@@ -0,0 +1,71 @@
+; Basic dump before and after a single module pass
+; RUN: opt %s -disable-output -passes='no-op-module' -ir-dump-directory %t/logs -dump-after=no-op-module -dump-before=no-op-module
+; RUN: find %t/logs -type f -print | FileCheck %s --check-prefix=SINGLE-RUN -DSOURCE_FILE_NAME=%s
+; SINGLE-RUN-DAG: [[SOURCE_FILE_NAME]]/0.NoOpModulePass.0-before.ll
```
The first `0` means this is the first pass to run at this nesting level. The second `0` means this is the first `NoOpModulePass` to run at this nesting level.

While dumping to files is less noisy, without this first number you lose the sequence in which passes run, which makes it harder to trace issues.

NuriAmari commented on llvm/test/Other/dump-before-after-filenames:
```diff
@@ -0,0 +1,71 @@
+; MULTIPLE-NESTED-RUNS-DAG: [[SOURCE_FILE_NAME]]/1.ModuleToFunctionPassAdaptor.0/1.PassManager<llvm::Function>.1/0.NoOpFunctionPass.0-before.ll
+; MULTIPLE-NESTED-RUNS-DAG: [[SOURCE_FILE_NAME]]/1.ModuleToFunctionPassAdaptor.0/1.PassManager<llvm::Function>.1/0.NoOpFunctionPass.0-after.ll
+; MULTIPLE-NESTED-RUNS-DAG: [[SOURCE_FILE_NAME]]/2.ModuleToFunctionPassAdaptor.1/1.PassManager<llvm::Function>.1/0.FunctionToLoopPassAdaptor.0/1.PassManager<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>.0/0.NoOpLoopPass.0-before.ll
+; MULTIPLE-NESTED-RUNS-DAG: [[SOURCE_FILE_NAME]]/2.ModuleToFunctionPassAdaptor.1/1.PassManager<llvm::Function>.1/0.FunctionToLoopPassAdaptor.0/1.PassManager<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>.0/0.NoOpLoopPass.0-after.ll
```
Surely a less verbose pass name than `PassManager<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>` is achievable, but I'm not sure how to best go about that.

NuriAmari replied:
I made an attempt at this, the trouble is the new PM instrumentation infrastructure only provides callbacks with the PassID, which is this verbose sometimes. The existing Class name to pass name mapping does not contain entries for all passes (ex. adaptors and PMs). 

I think more consistent names are doable, but perhaps something for another patch.

NuriAmari commented on llvm/test/Other/lit.local.cfg:
```diff
@@ -0,0 +1 @@
+config.suffixes = ['', '.ll']
```
This is because in LIT `%s` expands to the source file path with periods escaped like so: `.../dump-before-after-filenames\\.`, which won't match the output of `find`. 

I also considered running `ls` on each file, but I think the output of filecheck on find is a little easier to read when it fails.

asl commented on llvm/include/llvm/Passes/StandardInstrumentations.h:
```diff
@@ -69,6 +69,62 @@ class PrintIRInstrumentation {
+
+  // A table to store how many times a given pass has run at the current "nested
+  // level"
+  using PassRunsFrequencyTableT = DenseMap<StringRef, uint>;
```
Use standard types.
```suggestion
  using PassRunsFrequencyTableT = DenseMap<StringRef, unsigned>;
```

mshockwave commented on llvm/lib/Passes/StandardInstrumentations.cpp:
```diff
@@ -830,6 +831,182 @@ void PrintIRInstrumentation::registerCallbacks(
+}
+
+void DumpIRInstrumentation::dumpBeforePass(StringRef PassID, Any IR) {
+  if (isIgnored(PassID))
```
I think this function doesn't do any effect, maybe just leave it empty?

NuriAmari replied:
Which function do you mean?

mshockwave replied:
> Which function do you mean?

`DumpIRInstrumentation::dumpBeforePass`. Since both `isIgnored` and `shouldDumpBeforePass` do not have any side effect, I believe `DumpIRInstrumentation::dumpBeforePass` basically does nothing.

NuriAmari replied:
What about these lines:

```
SmallString<16> OutputPath =
      fetchCurrentInstrumentationDumpFile("-before.ll");
  std::error_code EC;
  llvm::raw_fd_ostream OS(OutputPath, EC, llvm::sys::fs::CD_CreateAlways);
  if (EC)
    report_fatal_error(Twine("Failed to open ") + OutputPath +
                       " to support dump-before: " + EC.message());

  OS << "*** IR Dump Before " << PassID << " on " << getIRName(IR) << " ***\n";
  unwrapAndPrint(OS, IR);
```

This method is responsible for printing the IR to a file before a pass runs.

mshockwave replied:
somehow I thought the brace on line 870 marks the end of this function, my bad. It's all good now

mshockwave commented on llvm/lib/Passes/StandardInstrumentations.cpp:
```diff
@@ -830,6 +831,182 @@ void PrintIRInstrumentation::registerCallbacks(
+  this->PIC = &PIC;
+
+  PIC.registerBeforeNonSkippedPassCallback(
+      [this](StringRef P, Any IR) { this->pushPass(P, IR); });
```
Could we omit `this` when possible? (ditto for other similar occurrences in this file)

NuriAmari replied:
I don't yet understand why, but `this` is not implicitly captured here, it fails to compile without an explicit capture.

mshockwave replied:
> I don't yet understand why, but `this` is not implicitly captured here, it fails to compile without an explicit capture.

Hmmm alright, then leave it as it for now I guess

mshockwave commented on llvm/include/llvm/Passes/StandardInstrumentations.h:
```diff
@@ -69,6 +69,62 @@ class PrintIRInstrumentation {
+  void dumpBeforePass(StringRef PassID, Any IR);
+  void dumpAfterPass(StringRef PassID, Any IR);
+
+  bool shouldDumpBeforePass(StringRef PassID);
```
Please mark these `shouldDump*` functions as `const`

mshockwave commented on llvm/include/llvm/Passes/StandardInstrumentations.h:
```diff
@@ -69,6 +69,62 @@ class PrintIRInstrumentation {
+
+  // A table to store how many times a given pass has run at the current "nested
+  // level"
+  using PassRunsFrequencyTableT = DenseMap<StringRef, unsigned>;
```
Is it possible to use `StringMap` here?

mshockwave commented on llvm/lib/Passes/StandardInstrumentations.cpp:
```diff
@@ -830,6 +831,182 @@ void PrintIRInstrumentation::registerCallbacks(
+  }
+
+  PassRunsFrequencyTableT &FreqTable = PipelineStateStack.back().FreqTable;
+  if (FreqTable.find(PassID) == FreqTable.end())
```
nit: `FreqTable.count(PassID)` will suffice

mshockwave commented on llvm/lib/Passes/StandardInstrumentations.cpp:
```diff
@@ -830,6 +831,182 @@ void PrintIRInstrumentation::registerCallbacks(
+
+  // Make sure the directory we wish to write our log file into exists.
+  StringRef ParentDirectory = sys::path::parent_path(OutputPath);
+  if (!ParentDirectory.empty()) {
```
coding style: the curly braces here could be removed

mshockwave reviewed (commented):
I like this feature, though I feel like in many cases `-print-changed` is good enough.
Also slightly tangent: I think LLVM's interim github PR policy is to discourage force push on PR branch in favor of amending commits: https://discourse.llvm.org/t/update-on-github-pull-requests/71540/53

NuriAmari commented:
> I like this feature, though I feel like in many cases `-print-changed` is good enough. 

I've never experienced it myself, but my teammate has said another reason to want this is that when using thinLTO, multiple modules are processed concurrently on different threads, and the printed output can end up interleaved and be hard to read.

> Also slightly tangent: I think LLVM's interim github PR policy is to discourage force push on PR branch in favor of amending commits: https://discourse.llvm.org/t/update-on-github-pull-requests/71540/53

Ah ok sorry I didn't know, will make new commits moving forward.

mshockwave commented:
> > I like this feature, though I feel like in many cases `-print-changed` is good enough.
> 
> I've never experienced it myself, but my teammate has said another reason to want this is that when using thinLTO, multiple modules are processed concurrently on different threads, and the printed output can end up interleaved and be hard to read.
> 

Ah that makes sense

Commit:
Fix tests to account for platform dependent output paths

The name of a particular pass is determined automatically
using the name of the type that implements in. This in turn
is determined using __PRETTY_FUNCTION__ which is not defined
in the standard and implementation specific.

This means the output paths differ slightly depending on the
platform.

To adjust tests for this:

- Add regexes that relax the requirements on output paths.
- Use find to find output paths without specifying
- Remove a test that was intended to test multiple modules being
  processing by a single pipeline, but was not really testing that

PR Feedback is also addressed.

NuriAmari committed:
Merge branch 'llvm:main' into dump-before-after

Commit:
Loosen filename regex in dump-before-afer-filenames

Tests are still failing in CI because the expected output
paths do not match. Loosen the regexes to allow for more
variation in the output paths.

NuriAmari commented:
I'm having a bit of a hard time getting the output path tests to pass in CI, while they pass on my Linux machine. I also can't see the output observed in CI, so its a little hard to fix.

I believe the variance comes from the output of this function being unreliable: https://github.com/llvm/llvm-project/blob/208fa9acc0ffe5a460bcd504229c24a8d3f87180/llvm/include/llvm/Support/TypeName.h#L27.

I've loosened the requirements on the paths again in the latest commit, hopefully that satisfies CI while still testing a reasonable amount.

Commit:
Fix Windows build

Use a relative source filename rather than an absolute source filename
when crafting an output path. Previously two absolute paths would be
concatenated like so:

```
/foo/bar + /baz/qux => /foo/bar/baz/qux
```

Which is fine on Linux, but on windows absolute paths start with
the partition (ie. C:\foo\bar) so we can't just append them.

NuriAmari commented:
Latest round of windows failures seems unrelated.

NuriAmari commented:
I've tried this on a real application, and the output paths quickly get extremely long. So long that we fail to create the directory structure. Perhaps the nesting idea isn't the right choice after all.

aeubanks commented:
What about an option that just redirects the PrintIRInstrumentation output to files instead of stderr? The files can be something like `$N-$modulename.ll` where `$N` is an int that increments.

NuriAmari commented:
> What about an option that just redirects the PrintIRInstrumentation output to files instead of stderr? The files can be something like `$N-$modulename.ll` where `$N` is an int that increments.

Yeah that's more or less what I'm working on now. My plan is:

- Add a flag that provides a directory to write files to
- If that flag is passed, write files of the structure `$N-$modulename-$passname-(before|after).ll` into that directory instead of writing to stderr
- If you pass both `-print-after=foo` and `-print-before=foo` the dump files for a particular run of Foo on a particular unit of IR should match except for the before / after suffix.

This should be relatively easy to implement in the old pass manager as well.
